### DISCUSSION TITLE
Use DataTemplate in a ContentControl

### DISCUSSION THREAD

**Discussion by SirZeeno** [AvaloniaUI/Avalonia#15349] [Q&A] [answered]
I have been trying to figure this out for two days, but don't seem to get far with it.
I am trying to create a Tab system and want the main page to change according to what tab is selected.
I have seen people use content controls on a simple level but never in conjunction with a tab system. 
Anything I have tried either resulted in me not being able to see anything get added to the ContentControl or everything being added but the carousel items that uses another DataTemplate. 
Before I tried using the tab and content control I did try and see if my approach works standalone, which it does, but doesn't anymore and I am lost on why. 

Any help is greatly appreciated!

repo: https://github.com/SirZeeno/Anime-Archive-Handler-GUI

most of the functionality lies within the MainView, TemplateClasses.cs, DailyFeatures.cs, and the App.axaml
Apologies in advance for the messy code, I am still figuring Avalonia out and haven't had much time for organization.

**Comment by thevortexcloud** [comment]
I have taken a fairly brief look at your code and I am not clear how you are setting your tab control content? Typically you just shove something into the tab item and it just works. EG 

```xaml
  <TabControl>
        <TabItem Header="1">
            <TextBlock Text="1 Content" />
        </TabItem>
        <TabItem Header="2">
            <ContentControl Content="{Binding VmViewModel}">
                <ContentControl.DataTemplates>
                    <DataTemplate DataType="vm:SecondVMViewModel">
                        <TextBlock Text="{Binding Text}" />
                    </DataTemplate>
                </ContentControl.DataTemplates>
            </ContentControl>
        </TabItem>
    </TabControl>
```

From the sound of things it sounds like you don't actually want a tab control, you just want a radio button that _looks_ like a tab.

**Reply by thevortexcloud** [reply, marked as answer]
I have hacked together an MVVM implementation using radio buttons. Although frankly this would be far simpler using a tab control. The code below is purely for demonstrative purposes, there are better ways to do it if you really want to do it this way. Which I would suggest you don't.

Main view XAML
```xaml
 <DockPanel>
        <ItemsControl ItemsSource="{Binding Pages}">
            <ItemsControl.ItemTemplate>
                <DataTemplate>
                    <RadioButton Content="{Binding Title}" GroupName="Pages" IsChecked="{Binding Active}" />
                </DataTemplate>
            </ItemsControl.ItemTemplate>
        </ItemsControl>
        <ContentControl DockPanel.Dock="Right" Content="{Binding CurrentPage}">
            <ContentControl.DataTemplates>
                <DataTemplate DataType="vm:PageBase">
                    <TextBlock Text="{Binding Title}" />
                </DataTemplate>
            </ContentControl.DataTemplates>
        </ContentControl>
    </DockPanel>
```

MainViewModel
```csharp

public class MainWindowViewModel : ViewModelBase {
    private PageBase _currentPage;
    public ObservableCollection<PageBase> Pages { get; } = new ObservableCollection<PageBase>();

    public PageBase CurrentPage {
        get => this._currentPage;
        set => this.RaiseAndSetIfChanged(ref this._currentPage, value);
    }

    public MainWindowViewModel() {
        this.Pages.Add(this.CreatePage("1"));
        this.Pages.Add(this.CreatePage("2"));
    }

    private PageBase CreatePage(string title) {
        var page = new PageBase() {
            Title = title,
        };

        page.PropertyChanged += (sender, args) => {
            if (args.PropertyName == nameof(PageBase.Active)) {
                this.RefreshCurrentPage();
            }
        };

        return page;
    }

    public void RefreshCurrentPage() {
        this.CurrentPage = this.Pages.FirstOrDefault(o => o.Active);
    }
```

PageBase

```csharp
    public class PageBase : ViewModelBase {
        private bool _active;
        public string Title { get; init; }

        public bool Active {
            get => this._active;
            set => this.RaiseAndSetIfChanged(ref this._active, value);
        }
    }
```

**Reply by SirZeeno** [reply]
yea. I am not setting up my content in the tab control, I am just setting it up outside the tab system and am more just reading the index of the tab to determine the change I want to make to the content at hand and display it on the content control in grid row 1 instead of 0 where the tab system is. what do you suggest I do though?

**Reply by thevortexcloud** [reply]
That's not really how you are meant to use tabs, that's why I said it sounded like you actually want radio buttons. Tabs are content controls themselves, which means they will render what ever the content of the active tab is. Using an external control to display the current tab content is superfluous. As I mentioned (and gave code for) you could use radio buttons to do that since radio buttons primarily set a true/false value, but at that point you may as well just use tabs.

Is there a particular reason you need to do it this way?

That said, if you did go with the radio button approach, you can restyle them to look like tabs if you want. Avalonia supports overriding the look of all built in controls.

**Reply by SirZeeno** [reply]
I don't have a particular reason to do it this way, I just looked in the documentation for Avalonia, saw that there was a tab system already in place, and saw that you can get the index or just display the content below it. so my thinking was that the index was there so it could be used to make changes outside the TabItems display in conjunction with the ContentControl. Thank you for helping me understand how it works though, I got pretty frustrated how it wasn't displaying properly.

Might I ask why you don't recommend using the tab control though when you say your code above would be simpler with it?

**Reply by timunie** [reply]
You can use TabStrip also. It's just I suggest to use the MVVM pattern tbh. Check also Avalonia.Samples

**Reply by thevortexcloud** [reply]
> index was there so it could be used to make changes outside the TabItems display

You can do this. You just need to listen for changes to one of the item/index properties (either via the events or by bindings). But in your case I would argue it does not make any sense to do it since you are not using the tab control as anything but a styled grouped button control. Which is exactly what a radio button is.

**Reply by thevortexcloud** [reply]
> Might I ask why you don't recommend using the tab control though when you say your code above would be simpler with it?

Sorry, I did not see your question. I said it's simpler because the tab control handles all the presentation for you. You don't have to make your own content presenter or manage what is currently being displayed (and by extension make sure everything is in sync). It will do it all for you. The control is strongly orientated around displaying content located in one of its children, which is basically what you are trying to do anyway. There is nothing really stopping you from putting your additional content inside the tab control as well as a bound control inside all of the tab items. Which would also be simpler than doing it using radio buttons. Similarly you can still have your main tab content in the tab control and use the selected item/index to update other things outside the control.

What you are trying to achieve in the way you are doing it can theoretically work using the tab control, but it's like trying to use a misshaped block of wood instead of a hammer to hammer a nail. Using the right tool in the right way will make the job significantly easier and more reliable.

**Reply by SirZeeno** [reply]
Fair enough, I am still trying to learn best practices for these situations because this is my first time trying my hand at Avalonia and I don't know the ins and outs of what I need to do yet. Thank you very much for the helping hand though, I appreciate it a lot.
